### PR TITLE
feat: structured type model (TypeRef) — phase 1 of the string-encoded-types replacement

### PR DESCRIPTION
**Pull request by ehabterra**
## What

Phase 1 of the biggest open architectural item (GAP #8): replace string-encoded types with a structured, typed model. This lands the foundation with **zero output drift**.

### `internal/typemodel` — one home for type-encoding knowledge

- **`TypeRef`** — a structured descriptor: package, name, generic args, and pointer/slice/array/map/chan constructors as *fields, not string fragments*. `Parse()` covers every encoding in use (`pkg-->Type`, dotted go/types form, wrapper prefixes, bracketed generics incl. nested/multi-arg/declaration-form, the legacy `pkg-->Type-->Arg` form) and never fails — unmodelable input stays opaque and renders back verbatim.
- **Renderers with round-trip guarantees** — `String()` (dotted), `Internal()` (`pkg-->Type[SimpleArgs]` component-key form), `Simple()` (unqualified, `interface{}`→`any`).
- **`FromExpr`** — the constructor at the AST/go-types boundary. A differential test in `internal/metadata` proves it **byte-identical to `getTypeName`** on every shape the legacy stringifier understands, and information-preserving where the flat string was lossy: multi-argument generic instantiations (`IndexListExpr` — previously stringified to `""`) and array lengths (previously rendered as slices).
- **`legacy.go`** — exact behavioral ports of the spec layer's string helpers (`ParseParts`, `NormalizeInstance`, `SimplifyArg`, `SimpleName`, `ArgPackage`, `SplitArgs`), each documenting its quirks and its structured replacement. `internal/spec` now delegates (`TypeParts` et al. are one-line wrappers; the duplicate parsers are deleted).

### Migration plan

`docs/TYPE_MODEL.md` lays out phases 2–3: migrate spec consumers onto `TypeRef` one PR at a time (each known legacy mangle becomes a deliberate, fixture-covered fix), then let metadata carry the structure itself. Ground rule going forward: new detection/resolution code builds/accepts a `TypeRef` instead of parsing type strings.

## Verification

- Full `go test ./...` green — golden metadata fixtures, determinism tests, and all framework fixture tests **unchanged** (zero drift).
- `-race` green on `typemodel`, `spec`, `metadata`, `generator`.
- `golangci-lint`, `go vet`, `gofmt` clean.
- New coverage: parse/render round-trip + structure tests, legacy-view pin tests (quirks asserted explicitly), and the `FromExpr`↔`getTypeName` boundary differential test.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Documentation**
  * Added guidance on type representation, compatibility expectations, and migration considerations.

* **Improvements**
  * Improved consistency when interpreting and displaying complex Go types, including generics, collections, channels, and qualified names.
  * Preserved compatibility with existing type-name formats while improving information retention.

* **Bug Fixes**
  * Reduced inconsistencies in type-name normalization and handling of nested generic arguments.

* **Tests**
  * Expanded coverage for parsing, rendering, round trips, edge cases, and compatibility with existing behavior.

### CODE REVIEW

File: internal/spec/extractor.go [hunk 1]
```diff
@@ -8,10 +8,13 @@ import (
...
 
 	"github.com/ehabterra/apispec/internal/metadata"
+	"github.com/ehabterra/apispec/internal/typemodel"
 )
 
 const (
-	TypeSep    = "-->"
+	// TypeSep aliases the type model's package/type separator; the single
+	// source of truth is internal/typemodel.
+	TypeSep    = typemodel.Sep
```
[the remark this answers is not in the document]
Comment (ehabterra) [reply internal/spec/extractor.go:17]: Not applicable: `TypeSep` is declared exactly once in the package (extractor.go). context_provider.go and mapper.go only *use* the constant — there is no declaration at their line 17 (verified with `grep -rn 'TypeSep\s*=' internal/spec`, and the package compiles: all CI builds are green).

File: internal/typemodel/legacy.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		parts.PkgName = baseParts[0]
+		parts.TypeName = baseParts[1]
+		// pkg-->Type-->Arg form (no brackets): trailing segments are the args.
+		if len(parts.GenericTypes) == 0 && len(baseParts) > 2 {
+			parts.GenericTypes = baseParts[2:]
+		}
+	} else if lastSep := strings.LastIndex(base, "."); lastSep > 0 {
+		parts.PkgName = base[:lastSep]
+		parts.TypeName = base[lastSep+1:]
+	} else {
```
[the remark this answers is not in the document]
Comment (ehabterra) [reply internal/typemodel/legacy.go:69]: This is an exact port of the original spec-layer TypeParts — the pre-existing implementation had the identical behavior for `Container[pkg.User]` (the bracket-peel gate requires a qualified base, then the dotted fallback scans the full string). Changing it here would violate this PR's byte-compatibility contract. Addressed by documenting it as a quirk in the ParseParts doc comment and pinning it with an explicit test case, so any future change is deliberate. In practice these forms are qualified by the canonicalizer before reaching lookups; the structured `Parse` handles them correctly for new code.

File: internal/typemodel/typemodel.go [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+		t.Elem = parse(s[len("chan "):])
+		return t
+	case strings.HasPrefix(s, "["):
+		// Array with a length: [N]Elem. Anything bracketed that doesn't
+		// close before more input falls through to the opaque named fallback.
+		if i := strings.IndexByte(s, ']'); i > 1 && i < len(s)-1 {
+			t.Kind = KindArray
+			t.Len = s[1:i]
+			t.Elem = parse(s[i+1:])
+			return t
```
[the remark this answers is not in the document]
Comment (ehabterra) [reply internal/typemodel/typemodel.go]: Fixed: array lengths now scan to the *matching* close bracket (new `matchingBracket` helper), so `[len([3]int{})]byte` parses and round-trips. Covered in TestRender.

File: internal/typemodel/typemodel.go [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+// list peeled off the end, then the base split into package qualifier and
+// simple name (internal Sep form or dotted form).
+func parseNamed(s string, t *TypeRef) {
+	t.Kind = KindNamed
+	base := s
+	if open := strings.Index(s, "["); open > 0 && strings.HasSuffix(s, "]") {
+		base = s[:open]
+		for _, a := range SplitArgs(s[open+1 : len(s)-1]) {
+			t.Args = append(t.Args, parseArg(a))
+		}
```
[the remark this answers is not in the document]
Comment (ehabterra) [reply internal/typemodel/typemodel.go:214]: Fixed: `SplitArgs` now tracks parenthesis depth as well, so a function-type argument (`F[func(int, string)]`) stays one argument. Covered in TestSplitArgs and TestRender. (The original splitGenericArgs had the same gap; since SplitArgs also feeds the structured parser, the fix strictly improves both layers — no fixture drift.)
